crypto/merklearray/merkle.go

```diff
@@ -17,63 +17,122 @@
 package merklearray
 
 import (
+	"bytes"
+	"errors"
 	"fmt"
+	"hash"
 	"sort"
 
 	"github.com/algorand/go-algorand/crypto"
 )
 
+const (
+	// MaxEncodedTreeDepth is the maximum tree depth (root only depth 0) for a tree which
+	// is being encoded (either by msbpack or by the fixed length encoding)
+	MaxEncodedTreeDepth = 16
+
+	// MaxNumLeavesOnEncodedTree is the maximum number of leaves allowed for a tree which
+	// is being encoded (either by msbpack or by the fixed length encoding)
+	MaxNumLeavesOnEncodedTree = 65536 // 2^MaxEncodedTreeDepth
+)
+
+// Merkle tree errors
+var (
+	ErrRootMismatch                  = errors.New("root mismatch")
+	ErrProvingZeroCommitment         = errors.New("proving in zero-length commitment")
+	ErrProofIsNil                    = errors.New("proof should not be nil")
+	ErrNonEmptyProofForEmptyElements = errors.New("non-empty proof for empty set of elements")
+	ErrUnexpectedTreeDepth           = errors.New("unexpected tree depth")
+	ErrPosOutOfBound                 = "pos %d larger than leaf count %d"
```

Comment:
the more idiomatic Go thing to do is to define this as an error type e.g. `errors.New("pos out of bound")` and then below wrap it with the specifics `fmt.Errorf("pos %d larger than leaf count %d: %w", idx, num, ErrPosOutOfBound)`. That way you can test it with errors.Is(err, ErrPosOutOfBound) as you are doing in your tests